subsys/bluetooth/host/gatt.c

```diff
@@ -1211,73 +1225,110 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
 }
 
-static void gatt_ccc_conn_unqueue(struct bt_conn *conn)
+static void gatt_delayed_store_dequeue(struct bt_conn *conn, enum delayed_store_flags flag)
 {
 	uint8_t index = bt_conn_index(conn);
+	struct conn_list *el = &gatt_delayed_store.conn_list[index];
 
-	if (gatt_ccc_store.conn_list[index] != NULL) {
-		bt_conn_unref(gatt_ccc_store.conn_list[index]);
-		gatt_ccc_store.conn_list[index] = NULL;
+	if (el->conn != NULL) {
+		__ASSERT_NO_MSG(el->conn == conn);
+		atomic_clear_bit(el->flags, flag);
+
+		if (atomic_get(el->flags) == 0) {
+			bt_conn_unref(el->conn);
+			el->conn = NULL;
+		}
 	}
 }
 
-static void gatt_ccc_conn_enqueue(struct bt_conn *conn)
+static void gatt_delayed_store_enqueue(struct bt_conn *conn, enum delayed_store_flags flag)
 {
-	if ((!gatt_ccc_conn_is_queued(conn)) &&
-	    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+	if (!conn) {
+		LOG_WRN("Not connected: cannot store client features to NVS");
+		return;
+	}
+
+	bool queued = gatt_delayed_store_is_queued(conn, flag);
+	bool bonded = bt_addr_le_is_bonded(conn->id, &conn->le.dst);
+
+	if (!queued && bonded) {
 		/* Store the connection with the same index it has in
 		 * the conns array
 		 */
-		gatt_ccc_store.conn_list[bt_conn_index(conn)] =
-			bt_conn_ref(conn);
+		uint8_t index = bt_conn_index(conn);
+		struct conn_list *el = &gatt_delayed_store.conn_list[index];
 
-		k_work_reschedule(&gatt_ccc_store.work, CCC_STORE_DELAY);
+		if (el->conn == NULL) {
+			el->conn = bt_conn_ref(conn);
+			atomic_set(el->flags, 0);
+		} else {
+			__ASSERT_NO_MSG(el->conn == conn);
+		}
+		atomic_set_bit(el->flags, flag);
+
+		k_work_reschedule(&gatt_delayed_store.work,
+				  K_MSEC(CONFIG_BT_SETTINGS_DELAYED_STORE_MS));
 	}
 }
 
-static bool gatt_ccc_conn_queue_is_empty(void)
+static bool gatt_delayed_work_queue_is_empty(void)
 {
 	for (size_t i = 0; i < CONFIG_BT_MAX_CONN; i++) {
-		if (gatt_ccc_store.conn_list[i]) {
+		if (gatt_delayed_store.conn_list[i].conn) {
 			return false;
 		}
 	}
 
 	return true;
 }
 
-static void ccc_delayed_store(struct k_work *work)
+static void delayed_store(struct k_work *work)
 {
 	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
-	struct gatt_ccc_store *ccc_store =
-		CONTAINER_OF(dwork, struct gatt_ccc_store, work);
+	struct gatt_delayed_store *store =
+		CONTAINER_OF(dwork, struct gatt_delayed_store, work);
 
 	for (size_t i = 0; i < CONFIG_BT_MAX_CONN; i++) {
-		struct bt_conn *conn = ccc_store->conn_list[i];
+		struct bt_conn *conn = store->conn_list[i].conn;
 
 		if (!conn) {
 			continue;
 		}
 
 		if (bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
-			ccc_store->conn_list[i] = NULL;
-			bt_gatt_store_ccc(conn->id, &conn->le.dst);
-			bt_conn_unref(conn);
+			if (IS_ENABLED(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE) &&
+			    gatt_delayed_store_is_queued(conn, DELAYED_STORE_CCC)) {
+				bt_gatt_store_ccc(conn->id, &conn->le.dst);
+				gatt_delayed_store_dequeue(conn, DELAYED_STORE_CCC);
```

Comment:
I think these effects must be done in reverse order. The flag must be cleared first. When the flag is set again while the store-procedure is in progress (possibly almost finished), it is necessary to run the procedure again.